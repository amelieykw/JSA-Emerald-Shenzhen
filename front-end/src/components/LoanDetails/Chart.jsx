import React, { useEffect, useState } from 'react'
import { VictoryChart, VictoryLine, VictoryScatter } from 'victory-native'
import { View } from 'react-native'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

const Chart = ({ loanHistory }) => {
  let mounted = true
  const [dataPair, setDataPair] = useState([])

  useEffect(() => {
    if (mounted && loanHistory) {
      let isOneYear = true
      const year = parseInt(loanHistory[0].date, 10)
      loanHistory.forEach(histItem => {
        if (year !== parseInt(histItem.date, 10)) {
          isOneYear = false
        }
      })
      if (isOneYear) {
        const dayDataPair = loanHistory.map(histItem => ({
          x: `${histItem.date.split('-')[1]}-${histItem.date.split('-')[2]}`,
          y: histItem.remaining
        }))
        setDataPair(dayDataPair)
      } else {
        const yearDataPair = []
        loanHistory.forEach(histItem => {
          if (yearDataPair.length === 0) {
            yearDataPair.push({
              x: histItem.date,
              y: histItem.remaining
            })
          } else if (
            parseInt(histItem.date, 10) !== parseInt(yearDataPair[yearDataPair.length - 1].x, 10)
          ) {
            yearDataPair.push({
              x: histItem.date,
              y: histItem.remaining
            })
          }
        })
        setDataPair(yearDataPair)
      }
    }
    return () => {
      mounted = false
    }
  }, [loanHistory])

  return dataPair.length > 0 ? (
    <View>
      <VictoryChart height={250}>
        <VictoryLine
          style={{ data: { stroke: '#FB7348' }, parent: { border: '1px solid #ccc' } }}
          data={dataPair}
        />
        <VictoryScatter data={dataPair} size={5} style={{ data: { fill: '#FB7348' } }} />
      </VictoryChart>
    </View>
  ) : null
}

Chart.propTypes = {
  loanHistory: PropTypes.arrayOf(PropTypes.any)
}

Chart.defaultProps = {
  loanHistory: undefined
}

const mapStateToProps = state => {
  return {
    loanHistory: state.loanList.find(loanItem => loanItem.id === state.detailLoanId).history
  }
}

export default connect(mapStateToProps, null)(Chart)
