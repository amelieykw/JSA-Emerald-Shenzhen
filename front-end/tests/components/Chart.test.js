import React from 'react'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'

import data from '../../helpers/mockData_FE'
import Chart from '../../src/components/LoanDetails/Chart'

const mockStore = configureStore([thunk])

describe('<Chart />', () => {
  it('<Chart /> should render correctly', () => {
    const store = mockStore({ loanList: data.detailedLoans, detailLoanId: 1 })
    const component = renderer.create(
      <Provider store={store}>
        <Chart />
      </Provider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
