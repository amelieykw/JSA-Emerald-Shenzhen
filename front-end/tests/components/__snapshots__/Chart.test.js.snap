// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> <Chart /> should render correctly 1`] = `null`;
